### PR TITLE
ESQL: Delay finding field load infrastructure

### PR DESCRIPTION
**Pull request by nik9000**
This optimizes loading fields across many, many indices by resolving the field loading infrastructure when it's first needed rather than up front. This speeds things up because, if you are loading from many many shards, you often don't need to set up the field loading infrastructure for all shards at all - often you'll just need to set it up for a couple of the shards.

### CODE REVIEW

**Comment by nik9000** [comment]
I've got some examples that this speeds up:
[no docs](https://gist.github.com/nik9000/bfd78dfa81e86db4b6e11d6450ba7867) 839ms -> 140ms
[docs](https://gist.github.com/nik9000/c127adc6accf4e8df9f3b44181aa7ff6) 1262ms -> 591ms

The change is much less substantial if there are docs, but it's still *something*.

**Comment by nik9000** [comment]
run  elasticsearch-ci/part-3

**Comment by dnhatn** [comment]
It's twice as fast in either case 👍

**Comment by nik9000** [comment]
> It's twice as fast in either case 👍

Yeah! I think `GuardedReader` was pretty expensive.

So this is one of the things we'd need before we could do #103666. We'd need to make the `BlockLoader`s have a real lifecycle to do that, but it felt like a good step. And it's faster!

**Review by costin — COMMENTED** [review]
Not a good reviewer - the only feedback I have is it moving the strings (like "long", "source_" ) to constants to avoid typos and subtle mistakes.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/ValuesSourceReaderBenchmark.java [hunk 1]
```diff
@@ -125,6 +126,26 @@ enum Where {
...
+        }
+        if (name.startsWith("keyword")) {
+            return ElementType.BYTES_REF;
+        }
+        throw new UnsupportedOperationException("no element type for [" + name + "]");
+    }
+
     private static BlockLoader blockLoader(String name) {
         Where where = Where.DOC_VALUES;
         if (name.startsWith("stored_")) {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/ValuesSourceReaderBenchmark.java:151]: Yeah, probably should. I'm not particularly careful in those benchmarks. I should be.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java [hunk 2]
```diff
@@ -226,50 +236,97 @@ public int get(int i) {
...
-                    StoredFieldsSpec storedFieldsSpec = storedFieldsSpecForShard(shard);
-                    LeafReaderContext ctx = ctx(shard, segment);
-                    storedFields = new BlockLoaderStoredFieldsFromLeafLoader(
-                        StoredFieldLoader.fromSpec(storedFieldsSpec).getLoader(ctx, null),
-                        storedFieldsSpec.requiresSource() ? shardContexts.get(shard).newSourceLoader.get().leaf(ctx.reader(), null) : null
-                    );
-                    if (false == storedFieldsSpec.equals(StoredFieldsSpec.NO_REQUIREMENTS)) {
-                        trackStoredFields(storedFieldsSpec, false);
+                    for (FieldWork field : fields) {
+                        field.clearAndMoveTo(shard, segment);
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:291]: It appears that we are not reloading the `FieldWork.loader` when the shard changes with `clearAndMoveTo`?

**Review by dnhatn — COMMENTED** [review]
One comment, but the PR looks great. Thanks Nik!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:291]: Oh darn. Let me look.That's bad. We totally should.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:291]: OK! I've reworked things a little and I think they are better.

**Comment by nik9000** [comment]
The performance after my last push doesn't really change. At least, not in a way I can detect with simple bash.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks, Nik!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java (hunk 2 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java]: Yes, the new method `positionFieldWork` is better.

**Pull request merged**